### PR TITLE
Destination Snowflake - add v1v2 migration

### PR DESCRIPTION
**Pull request by jbfbell**
- add snowflake migration scaffolding
- implement snowflake v1v2 migrator
- functional migration

## What
*Describe what the change is solving*
*It helps to add screenshots if it affects the frontend.*

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests#pull-request-title-convention)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  @SneakyThrows
+  @Override
+  protected boolean doesAirbyteInternalNamespaceExist(final StreamConfig streamConfig) {
+    return database
+        .queryJsons(
+            """
+                SELECT SCHEMA_NAME
+                FROM information_schema.schemata
+                WHERE schema_name = ?;
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java:37]: I think we also need to check `AND table_catalog = ?`, `databaseName`

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                """, streamConfig.id().rawNamespace())
+        .isEmpty();
+  }
+
+  @Override
+  protected boolean schemaMatchesExpectation(final SnowflakeTableDefinition existingTable, final Collection<String> columns) {
+    return CollectionUtils.containsAllIgnoreCase(existingTable.columns().keySet(), columns);
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java:44]: are we intentionally ignoring case here? (I think yes, this is the v1 table... but want to confirm)

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  @SneakyThrows
+  @Override
+  protected Optional<SnowflakeTableDefinition> getTableIfExists(final String namespace, final String tableName) {
+    // TODO this is mostly copied from SnowflakeDestinationHandler#findExistingTable, we should probably reuse this logic
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java:50]: :shrug:

this is all migration code, I imagine we intend to (a) never touch it again, and (b) delete it after X months

**Review by edgao — APPROVED** [review]
couple small comments but :shipit: after they're resolved

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java (hunk 3 above)
Comment (jbfbell) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java:50]: ha fair

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java (hunk 1 above)
Comment (jbfbell) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeV1V2Migrator.java:37]: good call, slightly different syntax of `AND catalog_name = ?` , `databaseName`. I also missed a `!` at the beginning of the expression 🤦

File: airbyte-integrations/bases/base-typing-deduping-test/src/main/resources/sqlgenerator/all_types_v1_inpurtrecords.jsonl
Comment (edgao) [inline airbyte-integrations/bases/base-typing-deduping-test/src/main/resources/sqlgenerator/all_types_v1_inpurtrecords.jsonl:1]: typo in the filename - `inpurt` -> `input`

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java [hunk 4]
```diff
@@ -212,26 +205,24 @@ protected void insertFinalTableRecords(final boolean includeCdcDeletedAt, final
             """));
   }
 
+  private String dollarQuoteWrap(JsonNode node) {
+    if (node == null) {
+      return "NULL";
+    }
+    final String stringContents = node.isTextual() ? node.asText() : node.toString();
+    // Use dollar quotes to avoid needing to escape anything
+    return StringUtils.wrap(stringContents, "$$");
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java:214]: meanwhile, in another PR https://github.com/airbytehq/airbyte/pull/29685/files#diff-f668de0cb12e6db0bb67714b015831f7df91559242f22a5d637b864c566dbc12R233

<.<

image

**Review by edgao — APPROVED** [review]

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java (hunk 4 above)
Comment (jbfbell) [reply airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java:214]: Should we change this to
```wrap(stringContents, "\\$\\$")```
?

File: airbyte-integrations/bases/base-typing-deduping-test/src/main/resources/sqlgenerator/all_types_v1_inpurtrecords.jsonl
Comment (jbfbell) [reply airbyte-integrations/bases/base-typing-deduping-test/src/main/resources/sqlgenerator/all_types_v1_inpurtrecords.jsonl:1]: :doh: but also lol

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java (hunk 4 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeSqlGeneratorIntegrationTest.java:214]: works for me!

**Pull request merged**